### ISSUE TITLE
Using X as dict with pipelines and scalers

### ISSUE THREAD

**Issue by blakewest** [skorch-dev/skorch#511]
I want to use the dict for the X variable of my pipeline. As part of my pipeline I use scalers like `MinMaxScaler`, or `RobustScaler`. These don't work with dictionary X values. As in.. the following would break...
```python
net = NeuralNetClassifier(module)
my_pipeline = [
  ('scaling', MinMaxScaler),
  ('classifier', net)
]
my_pipeline.fit({X: data, categorical_features: categorical_feature})
# error thrown because min max scaler doesn't know how to handle dict.
```
 I know there is this [SliceDict thing](https://nbviewer.jupyter.org/github/skorch-dev/skorch/blob/master/notebooks/Advanced_Usage.ipynb#Working-with-sklearn-FunctionTransformer-and-GridSearch), but I don't think that would work for my case, because it seems to only deal with the `len` issue, but my problem is deeper than that. I ended up creating a workaround wrapper class that looks like this...

```python
class MinMaxScalerForGroups(MinMaxScaler):
    ALL_DATA_KEY = 'X'

    def transform(self, X, y=None):
        if isinstance(X, dict):
            untransformed = X[self.ALL_DATA_KEY].copy()
            transformed = super().transform(X[self.ALL_DATA_KEY])
            for key in X:
                if key == self.ALL_DATA_KEY: continue

                transform = X[key].get("transform", None)
                column_indexes = X[key]["column_indexes"]
                print("Column indexes are:", column_indexes)
                if transform:
                    print("Using transformed values")
                    X[key] = transformed[:, column_indexes].copy()
                else:
                    print("Using untransformed values")
                    X[key] = untransformed.values[:, column_indexes].copy()

            X[self.ALL_DATA_KEY] = transformed
            return X
        else:
            return super().transform(X)

    def inverse_transform(self, X, y=None):
        if isinstance(X, dict):
            transformed = super().inverse_transform(X['X'])
            X[self.ALL_DATA_KEY] = transformed
            for key in X:
                column_indexes = X[key]
                X[key] = X[self.ALL_DATA_KEY].copy()[:, column_indexes]
            return X
        else:
            return super().inverse_transform(X)

    def fit(self, X, y=None):
        if isinstance(X, dict):
            return super().fit(X[self.ALL_DATA_KEY], y)
        else:
            return super().fit(X, y)
```

As you can see, it's pretty simple. It basically delegates to MinMaxScaler but handles the dict. Don't mind some of the goofy stuff around column_indexes. That's because I also wanted to specifically treat certain features differently in my loss function. I doubt this would be a mainstream use case. But my questions are...

1.) Is there a better way to deal with this?
2.) If not, I bet this could be made pretty generic as either a wrapper class that takes a scaler (or other sklearn class), and returns a version that works with dicts. Wanted to see if something like this might be useful to others. I could potentially make a PR

**Comment by BenjaminBossan** [comment]
It's true, sklearn cannot deal with dicts as input. We added it because for many use cases with neural networks, it is a good fit, but the disadvantage is that it won't play well with sklearn. As you mentioned, `SliceDict` is not a solution to this (as the name suggests, it only solves the slicing side of things).

When I run into similar issues as you did, I typically went with one of two solutions:

1. Change the data pipeline so that preprocessing is applied *before* the data is transformed into a dict.
2. Use [`dstoolbox`](https://github.com/ottogroup/dstoolbox).

With dstoolbox, you can do things like this:

```python
>>> import numpy as np
>>> from sklearn.pipeline import Pipeline
>>> from sklearn.preprocessing import StandardScaler
>>> from dstoolbox.pipeline import DictFeatureUnion
>>> from dstoolbox.transformers import ItemSelector

>>> X = {'a': np.arange(10).reshape(5, 2), 'b': np.arange(10, 20).reshape(5, 2)}
>>> X         
{
    'a': array([
        [0, 1],
        [2, 3],
        [4, 5],
        [6, 7],
        [8, 9]]),
    'b': array([
        [10, 11],
        [12, 13],
        [14, 15],
        [16, 17],
        [18, 19]])
}

>>> features = DictFeatureUnion([  # FeatureUnion that creates a dict
   ...:   ('a_scaled', Pipeline([  # 'a_scaled' will be the name of the new key
   ...:     ('select', ItemSelector('a')),  # select key 'a' from the dict
   ...:     ('scale', StandardScaler()),
   ...:   ])),
   ...:   ('b_scaled', Pipeline([  # 'b_scaled' will be the name of the new key
   ...:     ('select', ItemSelector('b')),  # select key 'b' from the dict
   ...:     ('scale', StandardScaler()),
   ...:   ])),
   ...: ])

>>> features.fit_transform(X)
{'a_scaled': array([
    [-1.41421356, -1.41421356],
    [-0.70710678, -0.70710678],
    [ 0.        ,  0.        ],
    [ 0.70710678,  0.70710678],
    [ 1.41421356,  1.41421356]]),
'b_scaled': array([
    [-1.41421356, -1.41421356],
    [-0.70710678, -0.70710678],
    [ 0.        ,  0.        ],
    [ 0.70710678,  0.70710678],
    [ 1.41421356,  1.41421356]])}

>>> model = Pipeline([('features', features), ('net', NeuralNet(...))])
```

There are many other goodies in that package, with some similar things being added slowly to sklearn.

However, dstoolbox is not a very popular package, so by all means use your own solution. That being said, I wouldn't add this kind of functionality to skorch. If you have transformers like this that have some very general application, consider contributing them to dstoolbox instead.

**Comment by BenjaminBossan** [comment]
@blakewest Any updates on this?

**Comment by blakewest** [comment]
@BenjaminBossan Not really. I guess I'll just keep with my current solution. I can't find it now, but I somewhere in your docs, you mention that skorch can take a `dict` as a param type. It might be good to just point out there that it won't really work in many pipelines. I could make a PR if you remember where that is discussed...

As for creating the dict after pre-processing, I'm not totally sure of an elegant way to do it for my use case, only because feature selection is part of my processing. I'd want to break up my features based on some kind of meaningful grouping, and after you apply a scaler, the names of the features go away. And after feature selections, the column ordering gets changed. So by the end of the pipeline, I don't have feature names or steady column indexes to use in order to take my non-dict features and turn them into a dict. There are obviously some dirty hacks like setting a global var, or otherwise trying to share state somehow, but that seems lame. Do you have other thoughts?

**Comment by BenjaminBossan** [comment]
I would probably be able to better help you if you showed your whole pipeline.

As I said earlier, you can achieve a lot with dstoolbox, e.g. with `ItemSelector`, you can choose a bunch of columns at the same time, and with `DataFrameFeatureUnion`, you can get back a DataFrame:

```python
>>> features = DataFrameFeatureUnion([  # FeatureUnion that works with dfs
   ...:   ('cardinal_features', Pipeline([  # scale all cardinal features
   ...:     ('select', ItemSelector(['col_a', 'col_b', 'col_c'])),  # select multiply columns
   ...:     ('scale', StandardScaler()),
   ...:     ('to_df', ToDataFrame(['col_a', 'col_b', 'col_c'])),  # transform back to df, keep original names
   ...:   ])),
   ...:   ('categorical_features', Pipeline([...])),
   ...: ])
```

Again, I think that your solution is perfectly valid, if it works for you, there is no point in using something else.
